### ISSUE TITLE
Newline Mode

### ISSUE THREAD

**Issue by shaunlebron** [parinfer/parinfer.js#98]
## tl;dr

> To follow the principle of least astonishment:
> 1. pressing enter should not alter the AST
> 2. pressing enter should not alter the position of the cursor in the AST
> 3. pressing enter and moving the cursor to another line should not alter the AST

A Newline Mode must be run (just once) after pressing Enter in Indent Mode.  This is because pressing "Enter" is an action that communicates intent, which cannot be implied from text alone.  This is inline with the purpose of Indent Mode and Paren Mode, which represent explicit user intentions.
## example of old behavior

Pressing enter at cursor below...

``` clj
(foo|) bar
```

...produces this text (before Indent Mode touches it):

``` clj
(foo
|) bar
```

After Indent Mode touches it, the close-paren is removed at start of the line...

``` clj
(foo
| bar
```

...and a close-paren is added after `bar` since it is now indented:

``` clj
(foo
| bar)
```
## new behavior

I'm proposing that the desired behavior produce the following output instead:

``` clj
(foo)
  |
bar
```

or the following with `previewCursorScope` enabled:

``` clj
(foo
  |)
bar
```

This is mainly to prevent `bar` from putting Parinfer into a bad state that is hard to recover from. _Full context on why and how below_...
## Background on Implicit Actions

Parinfer has a lot of implicit actions being performed by both modes.  I press `(` in Indent Mode, and this performs an implicit action of 'wrap'.  I indent the code in Indent Mode, and there is an implicit action of moving parens around. etc.

I say 'implicit' because this is not a keystroke-driven system where Parinfer responds to keys to perform its transformations.  Rather, it is the strings produced by the keystrokes which are the true catalyst of an action.  This is a subtle but important difference.
## Why 'Enter' cannot be implicitly performed

I'm mentioning this because I've realized that pressing 'Enter' is not an action that can be performed as an implicit reaction to the text it inserts, the 'newline' character.  Parinfer cannot just look at every newline in the same way.  Though it's tempting to view a newline character followed by a cursor as an implicit catalyst, it doesn't seem possible (reasons stated in next section).

That's why I'm considering 'Enter' to be the _first explicit action required by Parinfer_.  Technically, it is a third mode which takes text and produces new text.  But practically, it is only used when enter is pressed while in Indent Mode.
## Expectations when pressing Enter

Pressing enter is an implicit communication to Parinfer that the following Expectations are met:
1. a new line is inserted inside the current scope.
2. cursor should be positioned on the indentation point of the new line to signal the current scope.
3. the AST should remain the same if the new line is left blank

We cannot make number 2 an implicit action deduced from the proximity of a cursor to a newline character.  Otherwise, touching the start of a line with the cursor would cause the cursor to jump ahead to the indentation point.
## Newline Mode (one-time, not persistent)

Thus, I'm proposing that an editor running Parinfer in Indent Mode must run its text against a Newline Mode function after <kbd>Enter</kbd> is pressed.
## The Motivating Example

Suppose we start here in Indent Mode:

``` clj
(foo|) bar
```

Pressing enter gives the following text (before parinfer touches it):

``` clj
(foo
|) bar
```

We send the text to the Newline Mode function, which I am proposing should result in the following text:

``` clj
(foo)
  |
bar
```

(or the following if `previewCursorScope` is enabled)

``` clj
(foo
  |)
bar
```

Notice that:
1. inserted line is appropriately indented
2. `bar` was moved to the next line at the appropriate indentation level.
3. leaving the inserted line blank will not result in changes to the AST

**Comment by oakmac** [comment]
I'd like to see some more examples of this; my gut feel is that it will clash with existing Indent Mode rules and expected behavior.

It leaves the text in a state that Indent Mode would never create / allow. Are you proposing that some states are valid until "the next action" occurs (either with a cursor move or a key press)? If so, how will that work with text that is copy/pasted or loaded for the first time? How will that work with undo/redo?

In other words, this is a change that introduces the concept of state to Parinfer modes. How will this work with the stateless / pure function of Indent Mode?

**Comment by oakmac** [comment]
> Rather, it is the strings produced by the keystrokes which are the true catalyst of an action.

This is an absolutely brilliant insight, btw. It is the reason why Parinfer feels so natural compared to command-based modes like Paredit and command-based editors like Vi and Emacs.

Maybe a good candidate to add this insight to the main Parinfer website?

**Comment by oakmac** [comment]
I also wanted to reference the [section on Enter](https://shaunlebron.github.io/parinfer/#pressing-enter) on the main site. The change proposed in this Issue will be at odds with that section.

My gut feel is that <kbd>Enter</kbd> behavior should be handled by editor's autoindent rules; not Parinfer.

**Comment by shaunlebron** [comment]
> I'd like to see some more examples of this; my gut feel is that it will clash with existing Indent Mode rules and expected behavior.

To start with a non-problematic case:

``` clj
(foo [bar 1|])
```

After pressing enter, we have the following raw text which will be fed into Parinfer, assuming the editor auto-indented correctly:

``` clj
(foo [bar 1
      |])
```

If the new `previewCursorScope` option from #101 is enabled, the text will stay the same, but if you disable that, we will have the original behavior from the Enter section on the site:

``` clj
(foo [bar 1])
      |
```

That sums up the new cursor thing that I just finished, but let's look at a problematic case:

``` clj
(foo [bar 1|] baz)
```

When we press enter here, I'm proposing a new Enter function will transform to this:

``` clj
(foo [bar 1]
      |
  baz)
```

or this, if `previewCursorScope` is enabled:

``` clj
(foo [bar 1
      |]
  baz)
```

It doesn't matter which one is used; that's not the important part of this.  The important part is that:
- the new line will contain text that is in the _same structural place as the cursor before Enter was pressed_.
- `baz` maintains the same structural place.  It is just moved down a line so it doesn't create a very difficult state for Parinfer to recover from.

we can sync on this if it's still not clear

**Comment by shaunlebron** [comment]
> In other words, this is a change that introduces the concept of state to Parinfer modes. How will this work with the stateless / pure function of Indent Mode?

There is no state introduced that I can see.  The Enter function will produce code that doesn't change the AST, and applying either Mode function afterwards will not alter the AST either.  That is a very important property that is not satisfied if we allowed this:

``` clj
(foo [bar 1
      |] baz)
```

Hence, `baz` has to be pushed to the next line.

> Maybe a good candidate to add this insight to the main Parinfer website?

Thanks!  I'll think about where to put this.

**Comment by shaunlebron** [comment]
work happening in the [newline-mode branch](https://github.com/shaunlebron/parinfer/tree/newline-mode)

**Comment by shaunlebron** [comment]
restarted work on master: https://github.com/shaunlebron/parinfer/commit/9fc4b64e8d710751b006964bd3980978fa141078, but already noticing some required revisions based on the realization that this new transform must happen in two parts-- after pressing enter in Indent Mode:
1. AST stabilization - add extra whitespace to the cursor line
2. aligning sibling lines - ensure new line has same indentation as its siblings

Since item 2 is useful in both modes, I think I will just add a boolean `pressedEnter` option to both existing modes.  If enabled,  `indentMode` can enable a secondary `stabilizeNewline` option to perform the stabilizing split when necessary, and run `(processText ... PAREN_MODE)` first, then with `INDENT_MODE`.

This will result in removing the `newlineMode` function altogether and simpler editor usage `parinfer.indentMode(text, {pressedEnter: true, cursorX: ...})`

**Comment by shaunlebron** [comment]
Discovered that this problem is not particular to the enter key, but also applies to the backspace key.

```clj
;; input
(foo|) bar

;; action (enter)
(foo
  |) bar

;; Indent Mode (bad)
(foo
  bar)
```

```clj
;; input
(foo
  boo|) bar

;; action (backspace)
(foo
  |) bar

;; Indent Mode (bad)
(foo
  bar)
```

I think it's safe to discard this proposal now.  Trying something else here: https://github.com/shaunlebron/parinfer/blob/6422fc9977bd359e6e0c78590836983608588ca4/lib/doc/design.md#an-extra-rule
